Fix WCPay form fields references

cesarcosta99 opened the pull request:
Fixes #10193.

#### Changes proposed in this Pull Request

After #9940 the gateway `form_fields` is no longer initialized in `WC_Payment_Gateway_WCPay`. That changed made the call `$this->gateway->form_fields` in `WC_Payments_WooPay_Button_Handler` to evaluate to an empty array , causing the fatal error indicated in the issue.

We haven't been able to reproduce the issue live, but by doing a `var_dump($this->gateway->form_fields);` before [this condition](https://github.com/Automattic/woocommerce-payments/blob/573a5b3f75e1ef0bb2b18fc125e23c647255c091/includes/class-wc-payments-woopay-button-handler.php#L127) we can confirm `form_fields` is not evaluating as expected, and if a request ever gets to enter in that condition, it will get the fatal, which finally makes me conclude that's the root cause.

While searching for code references of `$this->gateway->form_fields` I noticed there's another in the settings REST controller, so I went ahead and changed it there too. This way we can prevent potential fatals on that class as well.

> [!WARNING]
> I'm aware this is going to conflict with #10236, but I deliberately created this separate PR to fix and test this issue in two fronts: a) by fixing the root cause in this PR; and b) by applying defensive programming in the other PR. I'm going to handle the conflicts in whichever gets merged last.

Related thread: p1737728531717659-slack-C01BZUL57SQ

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

**Reproduce the issue**

1. Checkout to `develop`.
2. Add `true ||` to [this condition](https://github.com/Automattic/woocommerce-payments/blob/573a5b3f75e1ef0bb2b18fc125e23c647255c091/includes/class-wc-payments-woopay-button-handler.php#L127):
```php
if (true || ! array_key_exists( self::BUTTON_LOCATIONS, get_option( 'woocommerce_woocommerce_payments_settings' ) ) ) {
```
3. Add a product to the cart and navigate to checkout.
4. The same fatal error indicated in the issue should be triggered:
`Fatal error: Uncaught TypeError: array_keys(): Argument #1 ($array) must be of type array, null given`

**Test: No fatal error is triggered**

1. Checkout to this branch.
2. Keep the condition change you did in step 2 above.
3. Refresh the checkout page.
4. You should not see any errors and page should load successfully.

**Regression test: WooPay selected pages is working as expected**

1. As a merchant, navigate to **Payments > Settings > Express checkouts > WooPay > Customize**.
2. Play with the **Enable WooPay button on selected pages** options by randomly checking and unchecking some of them.
3. As a customer, ensure the option is being saved and is working correctly by checking the WooPay express checkout button is being displayed in the pages you selected.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

cesarcosta99 committed:
Fix WCPay form fields reference in the WooPay button

cesarcosta99 committed:
Fix WCPay form fields reference in REST settings controller

cesarcosta99 committed:
Add changelog entry

cesarcosta99 commented:
@ricardo, I requested your review directly here once this is tied to #10236 and the testing is exactly the same with one addition.

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `10252` or branch name `fix/10193-form-fields-reference` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix/10193-form-fields-reference) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 7c24b63b4402e226f879c2e22ce4aa87b6b7ad0e
- Build time: 2025-01-29 19:43:26 UTC

_Note: the build is updated when a new commit is pushed to this PR._

ricardo commented:
@cesarcosta99 Do we have a way to pinpoint the changes from https://github.com/Automattic/woocommerce-payments/pull/9940 as a culprit for this error? I tried to `git revert 4608ebb` that commit from `develop` and added the `true ||`, but I could still reproduce the issue.

ricardo commented on includes/admin/class-wc-rest-payments-settings-controller.php:
```diff
@@ -945,12 +945,15 @@ private function update_woopay_enabled_locations( WP_REST_Request $request ) {
 
-		$all_locations = $this->wcpay_gateway->form_fields['payment_request_button_locations']['options'];
-		WC_Payments::woopay_tracker()->woopay_locations_updated( $all_locations, $woopay_enabled_locations );
+		if ( isset( $wcpay_form_fields['payment_request_button_locations']['options'] ) ) {
```
In this case, I'd be afraid this `if` statement could silence the original issue and create a different problem where the enabled locations settings are not updated.

I'd prefer only using `get_form_fields()` here as well, as we're not 100% sure what's causing the error, although I understand the intentional defensive programming.

cesarcosta99 replied:
As discussed in Slack, I reverted the added `if` since there are other calls to `get_form_fields` in the same file. Done in d99cace.

cesarcosta99 committed:
Revert if statement on settings REST controller

In order to avoid masking or silencing this and other failures, we're reverting the if statement added previously. Since no issues were reported so far, most likely this won't make a big difference.

Plus, there are other calls to  in the same file without extra checks, we should be on the safe side with this removal.

cesarcosta99 committed:
Merge branch 'develop' into fix/10193-form-fields-reference

cesarcosta99 commented:
> @cesarcosta99 Do we have a way to pinpoint the changes from #9940 as a culprit for this error?

@ricardo, yes, as I mentioned in p1737753080875879/1737728531.717659-slack-C01BZUL57SQ, you can start a git bisect at `develop` (preferably before #10236 to start with the bad behavior), checkout to say 8.5.0 (couple versions before 8.7.0) and use the condition change to test the issue. It should point you to #9940.

> I tried to `git revert 4608ebb` that commit from `develop` and added the `true ||`, but I could still reproduce the issue.

Considering the time you've left this comment, there's a chance you tested with #10236 merged in `develop`? If so, it'd prevent the fatal anyways.

At `develop`, I ran `git reset --hard 0454bb583` (immediate commit before #10236) and could reproduce the issue. Then, I did `git revert 4608ebb` (#9940), fixed the conflicts by accepting the incoming changes and was no longer able to reproduce the issue. Let me know if you're still having a different output.

ricardo commented on includes/class-wc-payments-woopay-button-handler.php:
```diff
@@ -126,7 +126,7 @@ public function init() {
 		if ( ! array_key_exists( self::BUTTON_LOCATIONS, get_option( 'woocommerce_woocommerce_payments_settings' ) ) ) {
 			if ( isset( $this->gateway->form_fields[ self::BUTTON_LOCATIONS ]['options'] ) ) {
-				$all_locations = $this->gateway->form_fields[ self::BUTTON_LOCATIONS ]['options'];
+				$all_locations = $this->gateway->get_form_fields()[ self::BUTTON_LOCATIONS ]['options'];
```
Since `$this->gateway->form_fields` wasn't available, we should probably use `->get_form_fields()` in the `isset` as well, what do you think? Or we might want to remove the `isset`?

cesarcosta99 replied:
Nice catch! I missed it when resolving the conflicts. Since we now have history of failure in this piece, I think it's better to keep the `isset`. Updated it in 7c24b63.

ricardo reviewed (commented):
@cesarcosta99 I was able to replicate the same issue after the revert. I must have mistakenly accepted the current changes instead of the incoming changes when fixing the conflict.

Just left a comment and I think we should be ready to merge.

cesarcosta99 committed:
Update if condition

ricardo reviewed (approved):
LGTM!
